virtinst/virtxml.py

```diff
@@ -181,35 +263,37 @@ def action_add_device(guest, options, parserclass, devs):
     return devs
 
 
-def action_remove_device(guest, options, parserclass):
-    if not parserclass.prop_is_list(guest):
-        fail(_("Cannot use --remove-device with --%s") %
-             parserclass.cli_arg_name)
-    if options.os_variant is not None:
-        fail(_("--os-variant/--osinfo is not supported with --remove-device"))
+def action_remove_device(action, guest):
+    parserclass = action.parserclass
+    parservalue = action.parservalue[-1]
 
     devs = _find_objects_to_edit(guest, "remove-device",
-        getattr(options, parserclass.cli_arg_name)[-1], parserclass)
+        parservalue, parserclass)
     devs = xmlutil.listify(devs)
 
     for dev in devs:
         guest.remove_device(dev)
     return devs
 
 
-def action_build_xml(options, parserclass, guest):
-    if not parserclass.guest_propname:
-        fail(_("--build-xml not supported for --%s") %
-             parserclass.cli_arg_name)
-    if options.os_variant is not None:
-        fail(_("--os-variant/--osinfo is not supported with --build-xml"))
+def action_build_xml(action, guest, options):
+    parserclass = action.parserclass
 
     devs = cli.run_parser(options, guest, parserclass)
     for dev in devs:
         dev.set_defaults(guest)
     return devs
 
 
+def perform_action(action, guest, options, input_devs):
+    if action.is_add_device:
+        return action_add_device(action, guest, options, input_devs)
+    if action.is_remove_device:
+        return action_remove_device(action, guest)
+    assert action.is_edit
+    return action_edit(action, guest, options)
```

Comment:
even if this is failsafe code, I'd use an approach similar to what done in `update_changes()`, ie
```python
if action.is_edit:
    return action_edit(action, guest, options)
raise xmlutil.DevError("perform_action(): unhandled action %s" % action.action_string)
```
(or `AssertionError` instead of `xmlutil.DevError`)

Reply:
all good points, I've addressed them now. thanks pino!